lib/sources/astroport-generator/index.ts

```diff
@@ -0,0 +1,102 @@
+import { SourceInterface } from '../../../types/source';
+import { Logger } from 'pino';
+import { Tendermint34Client } from '@cosmjs/tendermint-rpc';
+import { CbOnUserBalances } from '../../../types/cbOnUserBalances';
+import AstroportSource from '../astroport';
+import { queryContractOnHeight } from '../../query';
+
+export default class AstroportGeneratorSource implements SourceInterface {
+  rpc: string;
+  paginationLimit: number;
+  logger: Logger<never>;
+  generatorContract: string;
+  assets: Record<string, { denom: string; pair_contract: string }> = {};
+  sourceName: string;
+  client: Tendermint34Client | undefined;
+  astroportSource: AstroportSource;
+
+  getClient = async () => {
+    if (!this.client) {
+      this.client = await Tendermint34Client.connect(this.rpc);
+    }
+    return this.client;
+  };
+
+  constructor(rpc: string, logger: Logger<never>, params: any) {
+    this.logger = logger;
+
+    if (!params.source) {
+      throw new Error('No source name configured in params');
+    }
+    this.sourceName = params.source;
+
+    if (!params.assets) {
+      throw new Error('No assets configured in params');
+    }
+    this.assets = params.assets;
+
+    if (!params.generator_contract) {
+      throw new Error('No generator contract configured in params');
+    }
+    this.generatorContract = params.generator_contract;
+
+    this.rpc = rpc;
+    this.paginationLimit = parseInt(params.paginationLimit || '50', 10);
+    this.astroportSource = new AstroportSource(this.rpc, this.logger, params);
```

Comment:
I wouldn't use astroport source here as a field in class but a parent class. I mean it will be much clear to implement this class like this:
`class AstroportGeneratorSource implements AstroportSource`
Then, you will have direct access to all methods from a parent class and you can still redefine some

Reply:
That's what happens when you suddenly write OOP after many years in C, Go and Rust :laughing: 

Thank you!